### PR TITLE
[usage] Validate workspace instances in reconciler

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

Perform sanity checks on retrieved Workspace Instances. Excludes invalid instances and logs the problem with the records.

This is needed because our DB already contains records with instances which have poor data quality.

## Related Issue(s)

* Depends on https://github.com/gitpod-io/gitpod/pull/10490
* Depends on https://github.com/gitpod-io/gitpod/pull/10456

## How to test

Unit tests

## Release Notes

```release-note
NONE
```

## Documentation

NONE

### CODE REVIEW

**Comment by roboquat** [comment]
@easyCZ: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/usage/pkg/db/workspace_instance.go [hunk 1]
```diff
@@ -59,8 +59,9 @@ func ListWorkspaceInstancesInRange(ctx context.Context, conn *gorm.DB, from, to
 		Where(
 			conn.Where("stoppedTime >= ?", from).Or("stoppedTime = ?", ""),
 		).
-		Where("creationTime < ?", to).
-		Where("creationTime != ?", "").
```
Comment (easyCZ) [inline components/usage/pkg/db/workspace_instance.go:63]: This query would not match records without a creation time as `creationTime` cannot be both < `to` and `empty`

File: components/usage/pkg/controller/reconciler.go [hunk 2]
```diff
@@ -38,16 +47,100 @@ func (u *UsageReconciler) Reconcile() error {
...
 	if err != nil {
-		return fmt.Errorf("failed to list instances: %w", err)
+		return nil, fmt.Errorf("failed to load workspace instances: %w", err)
 	}
+	status.WorkspaceInstances = len(instances)
+	status.InvalidWorkspaceInstances = len(invalidInstances)
 
+	if len(invalidInstances) > 0 {
+		log.WithField("invalid_workspace_instances", invalidInstances).Errorf("Detected %d invalid instances. These will be skipped in the current run.", len(invalidInstances))
+	}
```
Comment (andrew-farries) [inline components/usage/pkg/controller/reconciler.go:73]: Do we want to have metrics on how many invalid workspaces we find?

File: components/usage/pkg/controller/reconciler.go [hunk 3]
```diff
@@ -38,16 +47,100 @@ func (u *UsageReconciler) Reconcile() error {
...
+
+func validateInstances(instances []db.WorkspaceInstance) (valid []db.WorkspaceInstance, invalid []invalidWorkspaceInstance) {
+	for _, i := range instances {
+		// i is a pointer to the current element, we need to assign it to ensure we're copying the value, not the current pointer.
+		instance := i
+
+		// Each instance must have a start time, without it, we do not have a baseline for usage computation.
+		if !instance.CreationTime.IsSet() {
+			invalid = append(invalid, invalidWorkspaceInstance{
+				reason:              "missing creation time",
```
Comment (andrew-farries) [inline components/usage/pkg/controller/reconciler.go:105]: Need constants for these literals?

**Review by andrew-farries — CHANGES_REQUESTED** [review]
Looks good, just a couple of questions

File: components/usage/pkg/controller/reconciler.go (hunk 2 above)
Comment (easyCZ) [reply components/usage/pkg/controller/reconciler.go:73]: Yep. Follow-up PRs.

File: components/usage/pkg/controller/reconciler.go (hunk 3 above)
Comment (easyCZ) [reply components/usage/pkg/controller/reconciler.go:105]: These are only used in a single place and intended for human consumption so they can fix it up. I think we're fine with inlined.

**Review by andrew-farries — APPROVED** [review]

**Pull request merged**